# Library and Configuration

In [1]:
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt

import kagglehub
import os
from pathlib import Path

from rapidfuzz import process

path = Path(kagglehub.dataset_download("olistbr/brazilian-ecommerce"))
print(os.listdir(path))

print("Library loaded!")

['olist_customers_dataset.csv', 'olist_geolocation_dataset.csv', 'olist_orders_dataset.csv', 'olist_order_items_dataset.csv', 'olist_order_payments_dataset.csv', 'olist_order_reviews_dataset.csv', 'olist_products_dataset.csv', 'olist_sellers_dataset.csv', 'product_category_name_translation.csv']
Library loaded!


# Load Data

In [2]:
df_customers      = pd.read_csv(path / "olist_customers_dataset.csv")
df_geolocation    = pd.read_csv(path / "olist_geolocation_dataset.csv")
df_orders         = pd.read_csv(path / "olist_orders_dataset.csv")
df_order_items    = pd.read_csv(path / "olist_order_items_dataset.csv")
df_order_payments = pd.read_csv(path / "olist_order_payments_dataset.csv")
df_order_reviews  = pd.read_csv(path / "olist_order_reviews_dataset.csv")
df_products       = pd.read_csv(path / "olist_products_dataset.csv")
df_sellers        = pd.read_csv(path / "olist_sellers_dataset.csv")
df_category_translation = pd.read_csv(
    path / "product_category_name_translation.csv"
)
print("Data Loaded!")

Data Loaded!


# Screening

## Customers

In [3]:
print(df_customers.head())

                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  


In [78]:
print(df_customers["customer_id"].nunique())
print(df_customers["customer_unique_id"].nunique())

99441
96096


In [4]:
df_customers["customer_city"].value_counts()

customer_city
sao paulo                   15540
rio de janeiro               6882
belo horizonte               2773
brasilia                     2131
curitiba                     1521
                            ...  
olhos d'agua                    1
pacotuba                        1
sao sebastiao do paraiba        1
benedito leite                  1
campos verdes                   1
Name: count, Length: 4119, dtype: int64

In [5]:
process.extract(
    "rio de janeiro",
    df_customers["customer_city"]
        .dropna().unique(),
    limit=20
)

[('rio de janeiro', 100.0, 10),
 ('cachoeiro de itapemirim', 85.5, 20),
 ('sao jose do rio preto', 85.5, 68),
 ('santa cruz do rio pardo', 85.5, 122),
 ('sao goncalo do rio abaixo', 85.5, 197),
 ('santa rosa de viterbo', 85.5, 208),
 ('rio negro', 85.5, 257),
 ('ferraz de vasconcelos', 85.5, 338),
 ('rio claro', 85.5, 360),
 ('sao jose do rio pardo', 85.5, 379),
 ('santo antonio de padua', 85.5, 406),
 ('santo antonio de jesus', 85.5, 487),
 ('alto paraiso de goias', 85.5, 512),
 ('rio verde', 85.5, 520),
 ('brejo da madre de deus', 85.5, 540),
 ('sao francisco de assis', 85.5, 630),
 ('sao luis de montes belos', 85.5, 746),
 ('barra de sao francisco', 85.5, 876),
 ('rio doce', 85.5, 900),
 ('santo antonio de posse', 85.5, 909)]

lokasi terlihat clean dengan kemungkinan perbedaan format penulisan kecil

## Geolocation 

In [6]:
print(df_geolocation.head())

   geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP  


In [7]:
df_geolocation["geolocation_city"].value_counts()

geolocation_city
sao paulo         135800
rio de janeiro     62151
belo horizonte     27805
são paulo          24918
curitiba           16593
                   ...  
tres irmaos            1
morro chato            1
valao do barro         1
jordanésia             1
são valentim           1
Name: count, Length: 8011, dtype: int64

jumlah geolokasi tidak sinkron dengan jumlah kota yang ada di brazil, akan dicek apakah ada kesalahan penulisan format

In [8]:
process.extract(
    "rio de janeiro",
    df_geolocation["geolocation_city"]
        .dropna().unique(),
    limit=20
)

[('rio de janeiro', 100.0, 1048),
 ('rio janeiro', 95.0, 1051),
 ('rio de janeiro, rio de janeiro, brasil', 90.0, 1049),
 ('ferraz de vasconcelos', 85.5, 50),
 ('rio claro', 85.5, 233),
 ('são josé do rio pardo', 85.5, 263),
 ('sao jose do rio pardo', 85.5, 264),
 ('santo antonio de posse', 85.5, 284),
 ('santo antônio de posse', 85.5, 286),
 ('santa rosa de viterbo', 85.5, 335),
 ('sao jose do rio preto', 85.5, 440),
 ('são josé do rio preto', 85.5, 441),
 ('aparecida de monte alto', 85.5, 630),
 ('bom sucesso de itarare', 85.5, 884),
 ('bom sucesso de itararé', 85.5, 886),
 ('aparecida de são manuel', 85.5, 906),
 ('aparecida de sao manuel', 85.5, 909),
 ('aguas de santa barbara', 85.5, 931),
 ('águas de santa bárbara', 85.5, 933),
 ('santa cruz do rio pardo', 85.5, 945)]

## Orders

In [9]:
print(df_orders.head())

                           order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08

In [87]:
valid_count = df_orders['customer_id'].isin(
    df_customers['customer_id']
).sum()

valid_count

np.int64(99441)

In [10]:
df_orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

order status valid

In [11]:
datetime_cols = df_orders.filter(like='date').columns.union(
    df_orders.filter(like='timestamp').columns.union(
    df_orders.filter(like='approved').columns)
)

before_na = df_orders[datetime_cols].isna().sum()

df_orders[datetime_cols] = df_orders[datetime_cols].apply(
    pd.to_datetime,
    format='%Y-%m-%d %H:%M:%S',
    errors='coerce'
)

after_na = df_orders[datetime_cols].isna().sum()

pd.DataFrame({
    'before_na': before_na,
    'after_na': after_na,
    'new_invalid': after_na - before_na
})

,before_na,after_na,new_invalid
order_approved_at,160,160,0
order_delivered_carrier_date,1783,1783,0
order_delivered_customer_date,2965,2965,0
order_estimated_delivery_date,0,0,0
order_purchase_timestamp,0,0,0


format order date dan timestamp konsisten

In [12]:
df_orders["order_id"].nunique() == len(df_orders)

True

valid bahwa order id bernilai unik

In [13]:
date_order_checks = {
    'approved_before_purchase':
        (df_orders['order_approved_at'] < df_orders['order_purchase_timestamp']).sum(),

    'carrier_before_approved':
        (df_orders['order_delivered_carrier_date'] < df_orders['order_approved_at']).sum(),

    'delivered_before_carrier':
        (df_orders['order_delivered_customer_date'] < df_orders['order_delivered_carrier_date']).sum(),

    'estimated_before_purchase':
        (df_orders['order_estimated_delivery_date'] < df_orders['order_purchase_timestamp']).sum(),

    'delivered_after_estimated':
        (df_orders['order_delivered_customer_date'] > df_orders['order_estimated_delivery_date']).sum(),
}

pd.Series(date_order_checks)

approved_before_purchase        0
carrier_before_approved      1359
delivered_before_carrier       23
estimated_before_purchase       0
delivered_after_estimated    7827
dtype: int64

In [14]:
(
    df_orders['order_approved_at']
    - df_orders['order_delivered_carrier_date']
).loc[
    df_orders['order_delivered_carrier_date']
    < df_orders['order_approved_at']
].describe()

count                         1359
mean     1 days 00:45:07.153053715
std      4 days 19:18:28.055754668
min                0 days 00:00:21
25%         0 days 01:24:55.500000
50%                0 days 17:10:04
75%                1 days 01:57:24
max              171 days 05:15:22
dtype: object

terdapat lebih dari 1000 baris yang memiliki urutan waktu dari tabel yang tidak sesuai, akan dijadikan validasi data

## Order Items

In [18]:
print(df_order_items.shape)
print(df_order_items.head())

(112650, 7)
                           order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99  

dataset order items berisikan order barang apa saja yang terjadi, setiap customer memesan barang, jumlah barang yang dipesan bisa lebih dari 1, maka dari itu jumlah baris order items lebih banyak dari df_orders itu sendiri

In [47]:
print(df_order_items["order_id"].nunique())
print(df_orders["order_id"].nunique())

98666
99441


In [38]:
df_orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

dengan selisih 775 dari dataset orders, for simplicity akan diasumsikan bahwa df orders hanya akan tercatat jika dan hanya jika barang akan sampai, walaupun 98441 - 98666 != number of canceled (625)

## Products

In [35]:
print(df_products.shape)
print(df_products.head())

(32951, 9)
                         product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            10

## Order reviews

In [49]:
print(df_order_reviews.shape)
print(df_order_reviews.head())

(99224, 7)
                          review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3

In [62]:
df_order_reviews.duplicated().sum()

np.int64(0)

In [88]:
print(df_order_reviews[
    df_order_reviews.duplicated(subset='review_id', keep=False)
].sort_values('review_id').head(6))

                              review_id                          order_id  \
46678  00130cbe1f9d422698c812ed8ded1919  dfcdfc43867d1c1381bfaf62d6b9c195   
29841  00130cbe1f9d422698c812ed8ded1919  04a28263e085d399c97ae49e0b477efa   
90677  0115633a9c298b6a98bcbe4eee75345f  78a4201f58af3463bdab842eea4bc801   
63193  0115633a9c298b6a98bcbe4eee75345f  0c9850b2c179c1ef60d2855e2751d1fa   
92876  0174caf0ee5964646040cd94e15ac95e  f93a732712407c02dce5dd5088d0f47b   
57280  0174caf0ee5964646040cd94e15ac95e  74db91e33b4e1fd865356c89a61abf1f   

       review_score review_comment_title  \
46678             1                  NaN   
29841             1                  NaN   
90677             5                  NaN   
63193             5                  NaN   
92876             1                  NaN   
57280             1                  NaN   

                                  review_comment_message review_creation_date  \
46678  O cartucho "original HP" 60XL não é reconhecid...  2018-03-07 0

## Order payments

In [90]:
print(df_order_payments.shape)
print(df_order_payments.head())

(103886, 5)
                           order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45  


In [111]:
(
    (df_order_payments[df_order_payments["order_id"] == "fa65dad1b0e818e3ccc5cb0e39231352"])
    .sort_values("payment_sequential")
)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
14321,fa65dad1b0e818e3ccc5cb0e39231352,1,voucher,1,3.71
23074,fa65dad1b0e818e3ccc5cb0e39231352,2,voucher,1,8.51
65641,fa65dad1b0e818e3ccc5cb0e39231352,3,voucher,1,2.95
9985,fa65dad1b0e818e3ccc5cb0e39231352,4,voucher,1,29.16
28330,fa65dad1b0e818e3ccc5cb0e39231352,5,voucher,1,0.66
29648,fa65dad1b0e818e3ccc5cb0e39231352,6,voucher,1,5.02
82593,fa65dad1b0e818e3ccc5cb0e39231352,7,voucher,1,0.32
68853,fa65dad1b0e818e3ccc5cb0e39231352,8,voucher,1,26.02
17274,fa65dad1b0e818e3ccc5cb0e39231352,9,voucher,1,1.08
19565,fa65dad1b0e818e3ccc5cb0e39231352,10,voucher,1,12.86


In [94]:
print(df_order_payments["order_id"].nunique())
print(df_order_payments["payment_sequential"].value_counts().sort_values())

99440
payment_sequential
29        1
28        1
27        1
24        2
25        2
26        2
23        2
22        3
21        4
20        4
17        6
16        6
19        6
18        6
15        8
14       10
13       13
12       21
11       29
10       34
9        43
8        54
7        82
6       118
5       170
4       278
3       581
2      3039
1     99360
Name: count, dtype: int64


In [95]:
check = (
    df_order_payments
    .groupby('order_id')['payment_sequential']
    .apply(lambda x: sorted(x.tolist()) == list(range(1, x.max()+1)))
)

check.value_counts()

payment_sequential
True     99360
False       80
Name: count, dtype: int64

In [96]:
invalid = (
    df_order_payments
    .groupby('order_id')['payment_sequential']
    .apply(lambda x: sorted(x.tolist()) != list(range(1, x.max()+1)))
)

invalid_orders = invalid[invalid].index

df_order_payments[
    df_order_payments['order_id'].isin(invalid_orders)
].sort_values(['order_id', 'payment_sequential'])

,order_id,payment_sequential,payment_type,payment_installments,payment_value
27191,00ac05fe0fc047c54418098eb64e3aaa,2,debit_card,1,123.47
12785,056c68d093c100017aab1f00f260705c,2,debit_card,1,52.78
40318,0668d086b3ae41adde3aed3dacbc8fae,2,debit_card,1,252.74
84637,0a7d898c6305101e69e9f5a05cd0130d,2,credit_card,5,215.47
82233,0d0e88a418636d40ea780339701db49c,2,debit_card,1,121.09
...,...,...,...,...,...
70969,f0f94b7c7548150f33f5d9e7597d396f,2,debit_card,1,148.95
49746,f17a58e669a7f4cf7a7f227899f8abf7,2,debit_card,1,230.15
6200,f5691c2b1ca263490374d13d020bd950,2,debit_card,1,172.98
15767,fce3db7092e1f432d0ec790414b6a5e4,2,debit_card,1,66.38


In [108]:
multi_cc = (
    df_order_payments
    .groupby('order_id')
    .filter(
        lambda x:
        (x['payment_type'] == 'credit_card').sum() > 1
    )
)

multi_cc.sort_values(['order_id', 'payment_sequential']).head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
101820,01358622f65717728b3b2ec2cf4e8812,1,credit_card,1,6.97
49201,01358622f65717728b3b2ec2cf4e8812,2,credit_card,1,14.80
99441,02d0cd05f04ab2e3c1acc473b0abc865,1,credit_card,3,56.40
79945,02d0cd05f04ab2e3c1acc473b0abc865,2,credit_card,2,29.64
11055,033ccfbdfc4d29677b7e1e6df3a82820,1,credit_card,3,30.82


In [107]:
df_order_payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

## Sellers

In [113]:
print(df_sellers.shape)
print(df_sellers.head())

(3095, 4)
                          seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP  


In [116]:
df_sellers["seller_city"].value_counts()

seller_city
sao paulo                                 694
curitiba                                  127
rio de janeiro                             96
belo horizonte                             68
ribeirao preto                             52
                                         ... 
ipua                                        1
muqui                                       1
timoteo                                     1
pouso alegre                                1
rio de janeiro, rio de janeiro, brasil      1
Name: count, Length: 611, dtype: int64

melalui value counts, kita sudah bisa melihat ada perbedaan penulisan di seller city, akan dinormalisasi nantinya

In [117]:
df_sellers["seller_state"].value_counts()

seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
CE      13
PE       9
PB       6
MS       5
RN       5
MT       4
RO       2
SE       2
AC       1
PI       1
MA       1
AM       1
PA       1
Name: count, dtype: int64

In [125]:
pd.read_csv(
    "https://raw.githubusercontent.com/datasets-br/city-codes/master/data/br-city-codes.csv",
    encoding="latin1"
)

,name,state,wdId,idIBGE,lexLabel,creation,extinction,postalCode_ranges,ddd,abbrev3,notes
0,Abadia de Goiás,GO,Q304652,5200050,abadia.goias,NaN,NaN,[75345-000 75349-999],62,NaN,NaN
1,Abadia dos Dourados,MG,Q582223,3100104,abadia.dourados,NaN,NaN,[38540-000 38549-999],34,ABD,NaN
2,Abadiânia,GO,Q304716,5200100,abadiania,NaN,NaN,[72940-000 72959-999],62,ABI,NaN
3,Abaeté,MG,Q1615444,3100203,abaete,NaN,NaN,[35620-000 35620-999],37,AET,NaN
4,Abaetetuba,PA,Q1615298,1500107,abaetetuba,NaN,NaN,[68440-000 68444-999],91,ABT,NaN
...,...,...,...,...,...,...,...,...,...,...,...
5565,Xique-Xique,BA,Q1762114,2933604,xique.xique,NaN,NaN,[47400-000 47439-999],74,XXQ,NaN
5566,Zabelê,PB,Q1808686,2517407,zabele,NaN,NaN,[58515-000 58519-999],83,NaN,NaN
5567,Zacarias,SP,Q1760645,3557154,zacarias,NaN,NaN,[15265-000 15269-999],18,ZAC,NaN
5568,Zé Doca,MA,Q1806488,2114007,ze.doca,NaN,NaN,[65365-000 65367-999],98,ZED,NaN


In [15]:
# df_customers      = pd.read_csv(path / "olist_customers_dataset.csv")
# df_geolocation    = pd.read_csv(path / "olist_geolocation_dataset.csv")
# df_orders         = pd.read_csv(path / "olist_orders_dataset.csv")
# df_order_items    = pd.read_csv(path / "olist_order_items_dataset.csv")
# df_order_payments = pd.read_csv(path / "olist_order_payments_dataset.csv")
# df_order_reviews  = pd.read_csv(path / "olist_order_reviews_dataset.csv")
# df_products       = pd.read_csv(path / "olist_products_dataset.csv")
# df_sellers        = pd.read_csv(path / "olist_sellers_dataset.csv")
# df_category_translation = pd.read_csv(
#     path / "product_category_name_translation.csv"
# )
# print("Data Loaded!")